File: android/core-imaging/src/main/java/com/offlinetracer/imaging/Color.kt
package com.offlinetracer.imaging

import kotlin.math.pow
import kotlin.math.sqrt

/**
 * Channels [Color.channel] can extract from an [RgbaImage].
 *
 * Declared at file level rather than nested inside [Color] so every consumer can spell it exactly
 * as the API contract writes it (`Channel.SATURATION`) with no import and no qualifier — the same
 * reason [GradientOp] lives at file level next to `Convolve`.
 */
enum class Channel { RED, GREEN, BLUE, ALPHA, LUMA, MAX, MIN, SATURATION, VALUE }

/**
 * Colour space conversions.
 *
 * Two luminances exist on purpose. [toGray] is Rec.601 luma on the *gamma-encoded* channels, which
 * is what OpenCV's `COLOR_RGB2GRAY` computes, so every threshold and sigma in the literature
 * transfers to this engine unchanged. [toGrayLinear] is Rec.709 luminance on *linearised* channels
 * and is used only by matting, where alpha is a linear quantity and mixing gamma-encoded values
 * into it produces visibly wrong edges on a composite.
 */
object Color {

    private const val INV255 = 1f / 255f

    /**
     * The same reciprocal in double, for the one expression that is required to be bit-identical to
     * the TypeScript engine's — see [toGray].
     */
    private const val INV255D = 1.0 / 255.0

    /**
     * sRGB transfer applied to the 256 possible 8-bit channel values.
     *
     * Built once because [linearize] costs a `pow` per channel: on a 12 MP image the straight
     * evaluation is 36 million `pow` calls, which measured slower than the entire rest of the
     * preprocessing chain.
     */
    private val SRGB_TO_LINEAR = FloatArray(256) { linearize(it * INV255) }

    // sRGB D65 primaries -> XYZ, and the D65 white point the Lab conversion normalises against.
    private const val XN = 0.95047f
    private const val YN = 1.00000f
    private const val ZN = 1.08883f

    // f(t) split point (6/29)^3 and the linear-segment coefficients 1/(3*(6/29)^2) and 4/29.
    private const val LAB_EPS = 0.008856452f
    private const val LAB_K = 7.787037f
    private const val LAB_OFF = 0.13793103f

    /**
     * Rec.601 luma of every pixel, `Y = 0.299R + 0.587G + 0.114B` on gamma-encoded channels,
     * scaled to 0..1. Alpha is ignored (not multiplied in): a transparent pixel keeps its colour,
     * because the matting stage owns the decision of what transparency means.
     *
     * **The weighted sum is evaluated in double and rounded to Float exactly once**, which is what
     * JavaScript does for free and what the TypeScript engine therefore already did. Float weights
     * round three times over and land up to 2 ulp (1.2e-7) away: measured on the `gradient-blob`
     * fixture, 186 of its 432 pixels differed. That is invisible against §14's 1e-4 on this stage and it
     * is *not* invisible downstream — it was the entire remaining 5.9e-6 residual in `EdgeDog.xdog`,
     * whose error gain is about a thousand (§7.2). Computing it identically in both engines is what
     * takes that stage to bit-identity, and it costs one narrowing per pixel.
     */
    fun toGray(src: RgbaImage): GrayF {
        val n = src.size
        val p = src.pixels
        val out = FloatArray(n)
        Parallel.chunks(n, Parallel.PIXELS_MAP) { from, to ->
            for (i in from until to) {
                val v = p[i]
                val r = (v ushr 16) and 0xFF
                val g = (v ushr 8) and 0xFF
                val b = v and 0xFF
                out[i] = ((0.299 * r + 0.587 * g + 0.114 * b) * INV255D).toFloat()
            }
        }
        return GrayF(src.width, src.height, out)
    }

    /**
     * Linear-light Rec.709 luminance, `Y = 0.2126R + 0.7152G + 0.0722B` on linearised channels.
     * The result is linear light in 0..1, deliberately *not* re-encoded to sRGB — the consumer
     * (matting) needs it linear.
     */
    fun toGrayLinear(src: RgbaImage): GrayF {
        val n = src.size
        val p = src.pixels
        val out = FloatArray(n)
        val lut = SRGB_TO_LINEAR
        Parallel.chunks(n, Parallel.PIXELS_MAP) { from, to ->
            for (i in from until to) {
                val v = p[i]
                val r = lut[(v ushr 16) and 0xFF]
                val g = lut[(v ushr 8) and 0xFF]
                val b = lut[v and 0xFF]
                out[i] = 0.2126f * r + 0.7152f * g + 0.0722f * b
            }
        }
        return GrayF(src.width, src.height, out)
    }

    /**
     * Grey to packed ARGB, values clamped and rounded to 0..255.
     *
     * With [opaque] false the grey doubles as the alpha channel, which is how an ink-density map or
     * a matte is previewed directly without a second buffer.
     */
    fun toRgba(src: GrayF, opaque: Boolean = true): RgbaImage {
        val n = src.size
        val s = src.data
        val out = IntArray(n)
        Parallel.chunks(n, Parallel.PIXELS_MAP) { from, to ->
            for (i in from until to) {
                val g = Px.toByte255(s[i])
                val a = if (opaque) 255 else g
                out[i] = (a shl 24) or (g shl 16) or (g shl 8) or g
            }
        }
        return RgbaImage(src.width, src.height, out)
    }

    /**
     * One channel of [src] as 0..1 floats. [Channel.MAX]/[Channel.VALUE] are the HSV value,
     * [Channel.SATURATION] is the HSV saturation `(max-min)/max` (0 for black, which has no hue).
     */
    fun channel(src: RgbaImage, channel: Channel): GrayF {
        val n = src.size
        val p = src.pixels
        val out = FloatArray(n)
        Parallel.chunks(n, Parallel.PIXELS_MAP) { from, to ->
            for (i in from until to) {
                val v = p[i]
                val a = (v ushr 24) and 0xFF
                val r = (v ushr 16) and 0xFF
                val g = (v ushr 8) and 0xFF
                val b = v and 0xFF
                out[i] = when (channel) {
                    Channel.RED -> r * INV255
                    Channel.GREEN -> g * INV255
                    Channel.BLUE -> b * INV255
                    Channel.ALPHA -> a * INV255
                    Channel.LUMA -> (0.299f * r + 0.587f * g + 0.114f * b) * INV255
                    Channel.MAX, Channel.VALUE -> {
                        val m = if (r > g) (if (r > b) r else b) else (if (g > b) g else b)
                        m * INV255
                    }
                    Channel.MIN -> {
                        val m = if (r < g) (if (r < b) r else b) else (if (g < b) g else b)
                        m * INV255
                    }
                    Channel.SATURATION -> {
                        val mx = if (r > g) (if (r > b) r else b) else (if (g > b) g else b)
                        val mn = if (r < g) (if (r < b) r else b) else (if (g < b) g else b)
                        if (mx == 0) 0f else (mx - mn).toFloat() / mx
                    }
                }
            }
        }
        return GrayF(src.width, src.height, out)
    }

    /** Alpha channel of [src] as 0..1 floats. */
    fun alphaOf(src: RgbaImage): GrayF {
        val n = src.size
        val p = src.pixels
        val out = FloatArray(n)
        for (i in 0 until n) out[i] = ((p[i] ushr 24) and 0xFF) * INV255
        return GrayF(src.width, src.height, out)
    }

    /**
     * Copy of [src] with its alpha replaced by [alpha] (0..1, clamped). RGB is untouched, i.e. the
     * result is non-premultiplied. Requires matching dimensions.
     */
    fun withAlpha(src: RgbaImage, alpha: GrayF): RgbaImage {
        require(src.width == alpha.width && src.height == alpha.height) {
            "withAlpha(): ${src.width}x${src.height} vs ${alpha.width}x${alpha.height}"
        }
        val n = src.size
        val p = src.pixels
        val a = alpha.data
        val out = IntArray(n)
        for (i in 0 until n) out[i] = (p[i] and 0x00FFFFFF) or (Px.toByte255(a[i]) shl 24)
        return RgbaImage(src.width, src.height, out)
    }

    /** sRGB transfer, gamma-encoded 0..1 to linear light. Values below the knee use the linear leg. */
    fun linearize(c: Float): Float =
        if (c <= 0.04045f) c / 12.92f else ((c + 0.055f) / 1.055f).pow(2.4f)

    /** Inverse of [linearize]: linear light to gamma-encoded sRGB. */
    fun delinearize(c: Float): Float =
        if (c <= 0.0031308f) 12.92f * c else 1.055f * c.pow(1f / 2.4f) - 0.055f

    /**
     * sRGB (gamma-encoded 0..1) to CIELAB D65, written into [out] as `L, a, b`.
     * [out] must have room for 3 floats; it is reused by the caller so this allocates nothing.
     */
    fun srgbToLab(r: Float, g: Float, b: Float, out: FloatArray) {
        require(out.size >= 3) { "srgbToLab(): out must hold 3 floats, got ${out.size}" }
        val rl = linearize(r)
        val gl = linearize(g)
        val bl = linearize(b)
        labFromLinear(rl, gl, bl, out, 0, 1, 2)
    }

    /** Plain ΔE76 — Euclidean distance in Lab. */
    fun labDistance(l1: Float, a1: Float, b1: Float, l2: Float, a2: Float, b2: Float): Float {
        val dl = l1 - l2
        val da = a1 - a2
        val db = b1 - b2
        return sqrt(dl * dl + da * da + db * db)
    }

    /**
     * Whole image as three planar Lab channels, `[L, a, b]`, each `width*height` in row-major order.
     * Planar rather than interleaved because every consumer (flood fill, magic wand) reads one
     * neighbour's three components at a stride of 1 image, and planes keep those three reads
     * cache-friendly across the scan.
     */
    fun toLabPlanes(src: RgbaImage): Array<FloatArray> {
        val n = src.size
        val p = src.pixels
        val l = FloatArray(n)
        val a = FloatArray(n)
        val b = FloatArray(n)
        val lut = SRGB_TO_LINEAR
        // Three `cbrt`s and a 3×3 matrix per pixel — the most expensive colour conversion in the engine
        // and the one matting waits on. The `tmp` triple is allocated **per share**: it exists to keep
        // the conversion allocation-free, and one triple shared between threads would hand a pixel
        // another pixel's `a` and `b`.
        Parallel.chunks(n, Parallel.PIXELS_MAP) { from, to ->
            val tmp = FloatArray(3)
            for (i in from until to) {
                val v = p[i]
                labFromLinear(
                    lut[(v ushr 16) and 0xFF],
                    lut[(v ushr 8) and 0xFF],
                    lut[v and 0xFF],
                    tmp, 0, 1, 2,
                )
                l[i] = tmp[0]
                a[i] = tmp[1]
                b[i] = tmp[2]
            }
        }
        return arrayOf(l, a, b)
    }

    private fun labFromLinear(
        rl: Float, gl: Float, bl: Float,
        out: FloatArray, li: Int, ai: Int, bi: Int,
    ) {
        val x = (0.4124564f * rl + 0.3575761f * gl + 0.1804375f * bl) / XN
        val y = (0.2126729f * rl + 0.7151522f * gl + 0.0721750f * bl) / YN
        val z = (0.0193339f * rl + 0.1191920f * gl + 0.9503041f * bl) / ZN
        val fx = labF(x)
        val fy = labF(y)
        val fz = labF(z)
        out[li] = 116f * fy - 16f
        out[ai] = 500f * (fx - fy)
        out[bi] = 200f * (fy - fz)
    }

    private fun labF(t: Float): Float =
        if (t > LAB_EPS) Math.cbrt(t.toDouble()).toFloat() else LAB_K * t + LAB_OFF
}

File: frontend/lib/trace/engine/color.ts
import { GrayF, RgbaImage, Px } from './buffers';

/**
 * Colour and luminance conversions. See ALGORITHMS.md §1.
 *
 * Two luminance definitions exist on purpose. {@link toGray} is Rec.601 luma on *gamma-encoded*
 * values, which is what OpenCV's `COLOR_RGB2GRAY` computes, so every threshold and sigma in the
 * literature transfers directly. {@link toGrayLinear} is physically-correct linear-light Rec.709 and
 * is used **only** by matting, where alpha is a linear quantity and the difference genuinely shows.
 */

/** Which scalar plane to pull out of an RGBA image. */
export enum Channel {
  RED = 'RED',
  GREEN = 'GREEN',
  BLUE = 'BLUE',
  ALPHA = 'ALPHA',
  LUMA = 'LUMA',
  MAX = 'MAX',
  MIN = 'MIN',
  SATURATION = 'SATURATION',
  VALUE = 'VALUE',
}

const INV255 = 1 / 255;

/**
 * Rec.601 luma of the gamma-encoded channels; alpha is ignored.
 * @returns a new GrayF with values in 0..1.
 */
export function toGray(src: RgbaImage): GrayF {
  const n = src.pixels.length;
  const out = new Float32Array(n);
  const px = src.pixels;
  for (let i = 0; i < n; i++) {
    const v = px[i];
    const r = (v >>> 16) & 0xff;
    const g = (v >>> 8) & 0xff;
    const b = v & 0xff;
    out[i] = (0.299 * r + 0.587 * g + 0.114 * b) * INV255;
  }
  return new GrayF(src.width, src.height, out);
}

/**
 * Linear-light Rec.709 luminance: channels are linearised first, then weighted 0.2126/0.7152/0.0722.
 * @returns a new GrayF with values in 0..1.
 */
export function toGrayLinear(src: RgbaImage): GrayF {
  const n = src.pixels.length;
  const out = new Float32Array(n);
  const px = src.pixels;
  // 256-entry LUT: linearize() is a pow() and doing it three times per pixel dominated this stage.
  const lut = linearizeLut();
  for (let i = 0; i < n; i++) {
    const v = px[i];
    out[i] =
      0.2126 * lut[(v >>> 16) & 0xff] + 0.7152 * lut[(v >>> 8) & 0xff] + 0.0722 * lut[v & 0xff];
  }
  return new GrayF(src.width, src.height, out);
}

let cachedLinearLut: Float32Array | null = null;

function linearizeLut(): Float32Array {
  let lut = cachedLinearLut;
  if (lut === null) {
    lut = new Float32Array(256);
    for (let i = 0; i < 256; i++) lut[i] = linearize(i * INV255);
    cachedLinearLut = lut;
  }
  return lut;
}

/**
 * Grey plane back to RGBA, values clamped to 0..1 on the way out.
 * @param opaque true writes alpha 255; false writes the grey level into alpha and leaves RGB black.
 */
export function toRgba(src: GrayF, opaque = true): RgbaImage {
  const n = src.data.length;
  const out = new Uint32Array(n);
  const d = src.data;
  if (opaque) {
    for (let i = 0; i < n; i++) {
      const b = Px.toByte255(d[i]);
      out[i] = (0xff000000 | (b << 16) | (b << 8) | b) >>> 0;
    }
  } else {
    // The grey doubles as alpha *and* stays in RGB. Zeroing RGB here instead would make an ink
    // density map preview as a black sheet at varying opacity rather than as the drawing it is.
    for (let i = 0; i < n; i++) {
      const b = Px.toByte255(d[i]);
      out[i] = ((b << 24) | (b << 16) | (b << 8) | b) >>> 0;
    }
  }
  return new RgbaImage(src.width, src.height, out);
}

/**
 * Extract one scalar plane. SATURATION and VALUE are the HSV definitions
 * (`s = (max-min)/max`, `v = max`); MAX/MIN are the raw channel extremes.
 * @returns a new GrayF with values in 0..1.
 */
export function channel(src: RgbaImage, ch: Channel): GrayF {
  const n = src.pixels.length;
  const out = new Float32Array(n);
  const px = src.pixels;
  for (let i = 0; i < n; i++) {
    const v = px[i];
    const a = (v >>> 24) & 0xff;
    const r = (v >>> 16) & 0xff;
    const g = (v >>> 8) & 0xff;
    const b = v & 0xff;
    let o: number;
    switch (ch) {
      case Channel.RED:
        o = r;
        break;
      case Channel.GREEN:
        o = g;
        break;
      case Channel.BLUE:
        o = b;
        break;
      case Channel.ALPHA:
        o = a;
        break;
      case Channel.LUMA:
        o = 0.299 * r + 0.587 * g + 0.114 * b;
        break;
      case Channel.MAX:
      case Channel.VALUE:
        o = r > g ? (r > b ? r : b) : g > b ? g : b;
        break;
      case Channel.MIN:
        o = r < g ? (r < b ? r : b) : g < b ? g : b;
        break;
      case Channel.SATURATION: {
        const mx = r > g ? (r > b ? r : b) : g > b ? g : b;
        const mn = r < g ? (r < b ? r : b) : g < b ? g : b;
        // mx == 0 is pure black, which has no defined hue or saturation; 0 is the conventional answer.
        out[i] = mx === 0 ? 0 : (mx - mn) / mx;
        continue;
      }
      default:
        throw new Error(`channel(): unknown channel ${String(ch)}`);
    }
    out[i] = o * INV255;
  }
  return new GrayF(src.width, src.height, out);
}

/** @returns the alpha plane scaled to 0..1. */
export function alphaOf(src: RgbaImage): GrayF {
  return channel(src, Channel.ALPHA);
}

/**
 * Replace the alpha plane. RGB is untouched (no premultiplication anywhere in this engine).
 * @throws if `alpha` is not the same size as `src`.
 */
export function withAlpha(src: RgbaImage, alpha: GrayF): RgbaImage {
  if (src.width !== alpha.width || src.height !== alpha.height) {
    throw new Error('withAlpha(): size mismatch');
  }
  const n = src.pixels.length;
  const out = new Uint32Array(n);
  const px = src.pixels;
  const a = alpha.data;
  for (let i = 0; i < n; i++) {
    out[i] = ((Px.toByte255(a[i]) << 24) | (px[i] & 0x00ffffff)) >>> 0;
  }
  return new RgbaImage(src.width, src.height, out);
}

/** sRGB transfer function, gamma-encoded 0..1 to linear-light 0..1. */
export function linearize(c: number): number {
  return c <= 0.04045 ? c / 12.92 : Math.pow((c + 0.055) / 1.055, 2.4);
}

/** Inverse sRGB transfer function, linear-light 0..1 to gamma-encoded 0..1. */
export function delinearize(l: number): number {
  return l <= 0.0031308 ? 12.92 * l : 1.055 * Math.pow(l, 1 / 2.4) - 0.055;
}

// D65 white point, matching the sRGB primaries below.
const XN = 0.95047;
const YN = 1.0;
const ZN = 1.08883;
const LAB_DELTA = 6 / 29;
const LAB_T0 = LAB_DELTA * LAB_DELTA * LAB_DELTA;
const LAB_M = 1 / (3 * LAB_DELTA * LAB_DELTA);

function labF(t: number): number {
  return t > LAB_T0 ? Math.cbrt(t) : LAB_M * t + 4 / 29;
}

/**
 * sRGB (gamma-encoded 0..1) to CIELAB D65, writing L,a,b into `out[0..2]`.
 * @param out must have length >= 3; written in place so per-pixel loops allocate nothing.
 */
export function srgbToLab(r: number, g: number, b: number, out: Float32Array | number[]): void {
  const rl = linearize(r);
  const gl = linearize(g);
  const bl = linearize(b);
  const x = 0.4124564 * rl + 0.3575761 * gl + 0.1804375 * bl;
  const y = 0.2126729 * rl + 0.7151522 * gl + 0.072175 * bl;
  const z = 0.0193339 * rl + 0.119192 * gl + 0.9503041 * bl;
  const fx = labF(x / XN);
  const fy = labF(y / YN);
  const fz = labF(z / ZN);
  out[0] = 116 * fy - 16;
  out[1] = 500 * (fx - fy);
  out[2] = 200 * (fy - fz);
}

/**
 * ΔE76 — plain Euclidean distance in Lab. ΔE2000 is not worth its cost here because every consumer
 * compares the result against a tolerance the user tunes by eye anyway.
 */
export function labDistance(
  l1: number,
  a1: number,
  b1: number,
  l2: number,
  a2: number,
  b2: number,
): number {
  const dl = l1 - l2;
  const da = a1 - a2;
  const db = b1 - b2;
  return Math.sqrt(dl * dl + da * da + db * db);
}

/**
 * Whole-image Lab conversion.
 * @returns `[L, a, b]`, each a `Float32Array` of `width * height`, in the same index order as the
 *          source pixels.
 */
export function toLabPlanes(src: RgbaImage): Float32Array[] {
  const n = src.pixels.length;
  const lp = new Float32Array(n);
  const ap = new Float32Array(n);
  const bp = new Float32Array(n);
  const px = src.pixels;
  const tmp = new Float32Array(3);
  for (let i = 0; i < n; i++) {
    const v = px[i];
    srgbToLab(
      ((v >>> 16) & 0xff) * INV255,
      ((v >>> 8) & 0xff) * INV255,
      (v & 0xff) * INV255,
      tmp,
    );
    lp[i] = tmp[0];
    ap[i] = tmp[1];
    bp[i] = tmp[2];
  }
  return [lp, ap, bp];
}
